File: Scraper/Filter.py
from pathlib import Path
from typing import Iterable

from infrastructure.models import Article

APPROVED_ISSNS = set()

with open(Path(__file__).parent/"assets/issns.txt") as f:
    APPROVED_ISSNS.update(line.strip() for line in f)


def filter_articles(articles: Iterable[Article]) -> Iterable[Article]:
    """
    Фильтрует список статей по ISSN, проверяя их наличие в списке журналов,
    утверждённых РЦНИ.
    """

    return filter(
        lambda article: article.journal and (
            article.journal.print_issn in APPROVED_ISSNS
            or article.journal.electronic_issn in APPROVED_ISSNS),
        articles)

File: TelegramBot/TelegramBot/Filter.cs
using Bot.Models;

namespace Bot.TelegramBot;

public static class Filter
{
    private static readonly HashSet<string> ApprovedIssns = new();

    static Filter()
    {
        var path = Path.Combine(Directory.GetCurrentDirectory(), "assets", "issns.txt");
        foreach (var line in File.ReadLines(path))
            ApprovedIssns.Add(line.Trim());
    }

    public static IEnumerable<Article> FilterArticles(IEnumerable<Article> articles) => articles.Where(article =>
        article.Journal is not null &&
        ((article.Journal.PrintIssn is not null && ApprovedIssns.Contains(article.Journal.PrintIssn)) ||
         (article.Journal.ElectronicIssn is not null && ApprovedIssns.Contains(article.Journal.ElectronicIssn))));
}